mchf-eclipse/drivers/audio/audio_driver.h

```diff
@@ -73,15 +73,42 @@
     #define NUM_AUDIO_CHANNELS 1
 #endif
 
+#ifdef USE_CONVOLUTION
+#define FFT_CONVOLUTION_SIZE 256
+#define CONVOLUTION_MAX_NO_OF_BLOCKS 8
+#define CONVOLUTION_MAX_NO_OF_COEFFS 2048
+#endif
+
 typedef struct
 {
     // Stereo buffers
     float32_t               i_buffer[IQ_BLOCK_SIZE];
     float32_t               q_buffer[IQ_BLOCK_SIZE];
 
-    float32_t               a_buffer[2][IQ_BLOCK_SIZE];
     float32_t               agc_valbuf[IQ_BLOCK_SIZE];   // holder for "running" AGC value
     float32_t               DF;
+
+#ifdef USE_CONVOLUTION
+    // for convolution filtering
+    int						nc; // no. of coefficients
+    int 					size; // no. of input samples
+    int						nfor; // no. of blocks in the convolution
+    int						buffidx; // buffer pointer
+    float32_t				impulse[CONVOLUTION_MAX_NO_OF_COEFFS * 2]; // impulse response has real and imaginary components
+    float32_t				i_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+    float32_t				q_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+    float32_t				maskgen[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fmask[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftin[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftout[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				ifftout[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				accum[FFT_CONVOLUTION_SIZE * 2];
+    float32_t               a_buffer[2][FFT_CONVOLUTION_SIZE / 2]; // for convolution, we need an output buffer of 128 samples
+#else
+    float32_t               a_buffer[2][IQ_BLOCK_SIZE];
+#endif
+
+    // for SAM demodulation
     float32_t               pll_fmax;
```

Comment:
I think it would be not a bad idea.  But read on:

BTW, all variables should be **defined** in audio_convolution.c only, all variables which will not be shared outside audio_convolution.h should get a static in front of them. And those shared with the audio_driver should be **declared** as extern in audio_convolution.h.
I.e. 
audio_convolution.h:
```
...
typedef struct
{
    float32_t				ifftout[FFT_CONVOLUTION_SIZE * 2];
    float32_t				accum[FFT_CONVOLUTION_SIZE * 2];
}  AudioConvolutionSharedVariables_t;
extern AudioConvolutionSharedVariables_t ac_shared; // declaration
...
```
In audio_convolution.c:

audio_convolution.h:
```
#include <audio_convolution.h>

AudioConvolutionSharedVariables_t ac_shared; // definition
...
typedef struct
{
    float32_t				a_buffer[FFT_CONVOLUTION_SIZE * 2];
}  AudioConvolutionNonSharedVariables_t;

static AudioConvolutionNonSharedVariables_t ac_nonshared;
...
```

You get the idea. Names are of course just examples...

Reply:
OK, thanks! got it
give me some time to digest now :-)

Reply:
How can I make the struct AudioSample_t available in convolution.h ?

Reply:
Yes,

move the typedef out of audio_driver.c (around line 110) into audio_driver.h 
and include audio_driver.h in audio_convolution.h before using AudioSample_t
.
Make sure to have no circular reference (i.e. audio_driver.h must not include audio_convolution.h) but this is should not be a problem.